docs/client-specifications/error-handling.md

```diff
@@ -1,3 +1,143 @@
+  CANCELLED_ERROR = 'CANCELLED_ERROR',                     // message: `The request was cancelled by the server; please contact Momento: ${detail}`
+  LIMIT_EXCEEDED_ERROR = 'LIMIT_EXCEEDED_ERROR',           // message: `Request rate exceeded the limits for this account.  To resolve this error, reduce your request rate, or contact Momento to request a limit increase.`
+  BAD_REQUEST_ERROR = 'BAD_REQUEST_ERROR',                 // message: `The request was invalid; please contact Momento: ${detail}`
+  TIMEOUT_ERROR = 'TIMEOUT_ERROR',                         // message: `The client's configured timeout was exceeded; you may need to use a Configuration with more lenient timeouts.  Timeout value: ${timeout}`
+  SERVER_UNAVAILABLE = 'SERVER_UNAVAILABLE',               // message: `The server was unable to handle the request; consider retrying.  If the error persists, please contact Momento.`
+  CLIENT_RESOURCE_EXHAUSTED = 'CLIENT_RESOURCE_EXHAUSTED', // message: `A client resource (most likely memory) was exhausted.  If you are executing a high volume of concurrent requests or using very large object sizes, your Configuration may need to be updated to allocate more memory.  Please contact Momento for assistance.`
+}
+```
+
+The `MomentoErrorTransportDetails` interface should capture the relevant gRPC information, for Momento debugging purposes:
+
+```typescript
+interface MomentoErrorTransportDetails {
+  grpc: MomentoGrpcErrorDetails
+}
+
+interface MomentoGrpcErrorDetails {
+  code: grpc.Status;
+  details: string;
+  metadata: grpc.Metadata;
+}
+```
+
+# Exposing Errors In Return Values
+
+**In languages that support sealed / algebraic data types, these should always be preferred.  In this document we are
+using TypeScript for our examples, so this illustrates more of a worst-case scenario for languages that lack such features.**
+
+`SimpleCacheClient` functions should return custom data types that are defined in the SDKs - they should not directly
+return generated gRPC response objects.
+
+## `CreateCache` and similar operations
+
+The `CreateCache` return value should have a `status` field, whose value is an enumeration including `SUCCESS` and `ERROR`.
+If the status is `ERROR`, then a `.error` field of type `Error` should be available:
+
+```typescript
+const result = client.createCache("myCache");
+if (result.status === CreateCacheStatus.SUCCESS) {
+  console.log(`Cache created`);
+} else if (result.status === CreateCacheStatus.ERROR) {
+  const error = result.error;
+  console.log(`An error occurred while attempting to create cache 'myCache': ${error.message}`);
+}
+```
+
+## `Get` and similar operations
+
+The `Get` return value should have a `status` field, whose value is an enumeration including `HIT`, `MISS` and `ERROR`.
+If the status is `ERROR`, then a `.error` field of type `Error` should be available:
```

Comment:
```suggestion
If the status is `ERROR`, then a `.error` field of type `Error` MUST be available:
```

I particularly like 1 aspect of RFC language, which is the caps for shoulds and musts to clarify the axes around which ambiguity lies and those where deviations are bugs.

Here & elsewhere, I just noticed how important this one is and I think it's a MUST, right?